include/pfn/expected.hpp

```diff
@@ -0,0 +1,1483 @@
+      noexcept(::std::is_nothrow_assignable_v<E, G const &> && ::std::is_nothrow_constructible_v<E, G const &>
+               && (::std::is_nothrow_move_constructible_v<E> || ::std::is_nothrow_move_constructible_v<T>)) // extension
+    requires(::std::is_constructible_v<E, G const &> && ::std::is_assignable_v<E &, G const &>
+             && (::std::is_nothrow_constructible_v<E, G const &> || ::std::is_nothrow_move_constructible_v<T>
+                 || ::std::is_nothrow_move_constructible_v<E>))
+  {
+    if (not set_) {
+      e_ = ::std::forward<G const &>(s.error());
+    } else {
+      _reinit(e_, v_, ::std::forward<G const &>(s.error()));
+      set_ = false;
+    }
+    return *this;
+  }
+
+  template <class G>
+  constexpr expected &operator=(unexpected<G> &&s) //
+      noexcept(::std::is_nothrow_assignable_v<E, G> && ::std::is_nothrow_constructible_v<E, G>
+               && (::std::is_nothrow_move_constructible_v<E> || ::std::is_nothrow_move_constructible_v<T>)) // extension
+    requires(::std::is_constructible_v<E, G> && ::std::is_assignable_v<E &, G>
+             && (::std::is_nothrow_constructible_v<E, G> || ::std::is_nothrow_move_constructible_v<T>
+                 || ::std::is_nothrow_move_constructible_v<E>))
+  {
+    if (not set_) {
+      e_ = ::std::forward<G>(s.error());
+    } else {
+      _reinit(e_, v_, ::std::forward<G>(s.error()));
+      set_ = false;
+    }
+    return *this;
+  }
+
+  template <class... Args>
+  constexpr T &emplace(Args &&...args) noexcept
+    requires(::std::is_nothrow_constructible_v<T, Args...>)
+  {
+    if (set_) {
+      ::std::destroy_at(::std::addressof(v_));
+    } else {
+      ::std::destroy_at(::std::addressof(e_));
+      set_ = true;
+    }
+    return *::std::construct_at(::std::addressof(v_), std::forward<Args>(args)...);
+  }
+
+  template <class U, class... Args>
+  constexpr T &emplace(::std::initializer_list<U> il, Args &&...args) noexcept
+    requires(::std::is_nothrow_constructible_v<T, ::std::initializer_list<U> &, Args...>)
+  {
+    if (set_) {
+      ::std::destroy_at(::std::addressof(v_));
+    } else {
+      ::std::destroy_at(::std::addressof(e_));
+      set_ = true;
+    }
+    return *::std::construct_at(::std::addressof(v_), il, std::forward<Args>(args)...);
+  }
+
+  // [expected.object.swap], swap
+  constexpr void
+  swap(expected &rhs) noexcept(::std::is_nothrow_move_constructible_v<T> && ::std::is_nothrow_swappable_v<T>
+                               && ::std::is_nothrow_move_constructible_v<E> && ::std::is_nothrow_swappable_v<E>)
+    requires(::std::is_swappable_v<T> && ::std::is_swappable_v<E> && ::std::is_move_constructible_v<T>
+             && ::std::is_move_constructible_v<E>
+             && (::std::is_nothrow_move_constructible_v<T> || ::std::is_nothrow_move_constructible_v<E>))
+  {
+    bool const lhset = has_value();
+    bool const rhset = rhs.has_value();
+    if (lhset == rhset) {
+      if (lhset) {
+        using ::std::swap;
```

Comment:
Could probably do this once one `if` higher?

Reply:
This is literally what the C++ standard prescribes, see table 72 (left top) at https://eel.is/c++draft/expected.object.swap

Reply:
Are you suggesting that there is a difference on which line "using" appears? Would it not be the same if you put the "using" only once but above the "if/else"?